### ISSUE TITLE
Output ES Modules

### ISSUE THREAD

**Issue by dandv** [typegoose/typegoose#214]
I'm targeting `esnext` in tsconfig, and trying to `import { prop } from '@typegoose/typegoose';` fails with 

> SyntaxError: The requested module '@typegoose/typegoose' does not provide an export named 'prop'

Same issue that [graphql-tools had](https://github.com/apollographql/graphql-tools/issues/913).

**Comment by hasezoey** [comment]
sorry, but when doing that, it might affect older versions of node and might affect the code of the users

**Comment by dandv** [comment]
Not if released in a new major version.
>
>

**Comment by dandv** [comment]
> sorry, but when doing that, it might affect older versions of node and might affect the code of the users

I doubt it. Much bigger projects, with many more users, like `graphql-tools` which I linked above, have done it.

Also, [graphql-modules agreed to implement ES Modules](https://github.com/Urigo/graphql-modules/issues/1066#issuecomment-612544335).

Anyway, there's an easy way to prevent any incompatibilities: release a new major version announcing the *addition* of ES Modules. That way, npm won't automatically update the user's typegoose package beyond the current version.

**Comment by hasezoey** [comment]
i have looked more into it, i would like to emit ES-Modules, but they are only included in node 13 (without flag) and from what i know are still somewhat experimental and the next version 7.0 will remove nodejs 8 support (node 10 will be the lowest supported then), so for the near future this will probably not be happening - sorry

PS: i will look into this again when the time comes

**Comment by dandv** [comment]
Maybe the release of Node 14 is a good reason to look into this. I ended up using a nasty workaround to get typegoose to play with `"modules": "esnext"`, which apparently [breaks TypeScript reflection](https://github.com/typegoose/typegoose/pull/245#issuecomment-626119383).

```ts
import * as typegoose from '@typegoose/typegoose';
import { DocumentType, Ref } from '@typegoose/typegoose';  // these are types
const { index, prop, arrayProp, getModelForClass, defaultClasses } = typegoose['default'] || typegoose;
```

**Comment by 101arrowz** [comment]
This actually brings up a bigger question; since Typegoose is targeted solely towards TypeScript users, why not remove the build step entirely and export the raw `.ts` files to NPM? The user then gets to decide whether to use CommonJS or ES Modules for the final output.

**Comment by hasezoey** [comment]
@101arrowz you cant really import and compile ts files installed from an provider (like npm), at least i dont know and simple way

**Comment by dandv** [comment]
This may work for other reasons (Gatsby uses TypeScript) but here's an example of an NPM package that only ships `.ts` files: https://github.com/epilande/gatsby-theme-auth0/blob/master/package.json

**Comment by andreialecu** [comment]
I think simply adding a bunch of `.mjs` files to the npm repository can't hurt anything. Also see: https://nodejs.org/api/esm.html#esm_conditional_exports

Everything should stay backwards compatible even if modules were supported.

**Comment by hasezoey** [comment]
i tried to add esm and test it, but i couldnt get it to work, while also including cjs
branch: https://github.com/typegoose/typegoose/tree/feature/esm

run `npm run build` and it will output `/lib as cjs` and `/lib/esm as mjs`

i loaded it into my demo project
with
tsconfig
```json
{
  "compilerOptions": {
        "target": "es6",
        "module": "ES6",
        "allowSyntheticDefaultImports": true
  }
}
```
and package
```json
{
  "type": "module"
}
```

and code
```ts
// NodeJS: 14.7.0
// MongoDB: 4.2-bionic (Docker)
import { modelOptions, prop, getModelForClass } from "@typegoose/typegoose/lib/esm";
import mongoose from "mongoose"; // mongoose@5.9.26 @types/mongoose@5.7.34

@modelOptions({ schemaOptions: { timestamps: true, strict: "throw", collection: "user" } })
class User {
  @prop({ required: true, index: true })
  public userId!: number;
}
const UserModel = getModelForClass(User);

(async () => {
  await mongoose.connect(`mongodb://localhost:27017/`, { useNewUrlParser: true, dbName: "esm", useCreateIndex: true, useUnifiedTopology: true });



  await mongoose.disconnect();
})();
```

([typegoose-testing/esm](https://github.com/typegoose/typegoose-testing/tree/esm))

and got error `Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/home/hasezoey/Downloads/typegoose-testing/node_modules/@typegoose/typegoose/lib/esm/logSettings' imported from /home/hasezoey/Downloads/typegoose-testing/node_modules/@typegoose/typegoose/lib/esm/typegoose.js`
then tried to replace
```ts
import { logger } from './logSettings';
// with
import { logger } from 'logSettings';
```
and the next module came up - but i cant ship it this was and i didnt find an typescript option to export it the correct way

if someone can fix this and make an pr, i would merge it

**Comment by dandv** [comment]
I've just published a dual package using conditional exports. It can be imported by name from ESM TypeScript code (`"module": "esnext"`).

https://github.com/dandv/local-iso-dt.

Here is [the diff](https://github.com/dandv/local-iso-dt/commit/00d02af5fcaf06ca17c3f2b9e86f44ddc174b628).

**Comment by RyannGalea** [comment]
Any news on this, facing the same issue as @hasezoey example above when trying to do a prod build

**Comment by hasezoey** [comment]
this issue might be worth reconsidering (i tried some times before) with [typescript 4.5](https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/) because from then on it allows to use ESM way more conveniently

**Comment by AndyClausen** [comment]
I solved a lot of issues with esm imports by setting this node flag: `--experimental-specifier-resolution=node` - maybe this is useful to you as well.

**Comment by waynehaffenden** [comment]
Any updates on when this is gonna get merged and published?

**Comment by hasezoey** [comment]
> Any updates on when this is gonna get merged and published?

no, no updates, last time i tried i didnt get typescript to cooperate, but will try again when upgrading to typescript 4.5 (see #644)

**Comment by willster277** [comment]
I cloned this repo and made no changes other than
```diff
-"module": "commonjs",
+"module": "esnext",
```
(I didn't update any packages.)
Then ran 
```sh
npm run build:tests
npm test
```
and got the output
```
Test Suites: 24 passed, 24 total
Tests:       255 passed, 255 total
Snapshots:   67 passed, 67 total
Time:        40.246 s
Ran all test suites.
```
I also ran
```sh
npm run build
```
which completed successfully and with no errors.


I then updated typescript to the latest release with
```
npm i -D typescript
```
and ran the same commands again, I got the same successful results.

---

Currently at work, we've been pretty much backed into a corner (by updates to services we can't control) where the only options are to get hold of an ESM version of typegoose, or rewrite everything in mongoose or plain mongoDB. 
We could fork typegoose and maintain our own ESM version util this one upgrades to ESM, but none of us particularly want to maintain a package this large.

Is v10 going to be released any time soon?
ESM is now a non-negotiable necessity for us, and for anyone using a similar enough setup to ours.

**Comment by RyannGalea** [comment]
I noticed this ORM on Twitter the other day, I have not spent any time looking deep into it yet but it's compatible with MongoDB & might be an avenue to look into, just thought I'd share

https://github.com/mikro-orm

**Comment by hasezoey** [comment]
@willster277 
> Is v10 going to be released any time soon?

currently no major features are planned, so no

> which completed successfully and with no errors.

like i said earlier, this is how far i also got, but i got problems providing *both* at the same time, which is what i would like to do before completely migrating 

>  ESM is now a non-negotiable necessity for us, and for anyone using a similar enough setup to ours.

is ESM really required and not working with CJS? also, even if typegoose moves, mongoose and mongodb (from what i know) have not moved yet and also do not provide any ESM

**Comment by hasezoey** [comment]
> You could just build the codebase twice, once with ESM and once with CJS, and provide both exports.

that is what i wanted to do

> I've done this many times before, you just need to specify multiple fields in your package.json

i have also got as far as that in my tries, but it somehow never worked, either it was only providing the default, or it was providing nothing so i have up and wanted to try again later

Note: i was also trying at the same type to have the typings be correct and provided (i didnt care if i needed 1 or 2 typing files, but typescript never wanted to do it), so with typescript 4.5, the types should also work better and compile should also work better, i will re-investigate this when upgrading the typescript version next time i do it

**Comment by hasezoey** [comment]
i have updated the `feature/esm` branch, this time it compiled without problems, though i still have problems importing it - maybe someone can help out (this will not be included until this can be resolved and documented)

Current Problems:
- i cannot import ESM modules into a ESM project, error `src/test.ts:3:10 - error TS2305: Module '"@typegoose/typegoose"' has no exported member 'prop'` (typescript error, not runtime)
- i cannot import the CJS modules into a CJS project, error `src/test.ts:3:28 - error TS2305: Module '"@typegoose/typegoose"' has no exported member 'prop'.` (typescript error, not runtime)

project for esm testing: https://github.com/typegoose/typegoose-testing/tree/test_esm
project for cjs testing: https://github.com/typegoose/typegoose-testing/tree/master

to replace typegoose:
- compile typegoose with `yarn run build`
- run `rm -rf /typegoose-testing/node_modules/@ŧypegoose/typegoose`
- run `cp -r /typegoose/lib /typegoose-testing/node_modules/@ŧypegoose/typegoose/*`
Note: replace `/`(root) with the path to the mentioned projects

**Comment by willster277** [comment]
Until you're ready to fully switch to ESM, a possibility might be to have an `@esm` tag?

The build process might look like this (bash);
```sh
# get version from package.json (run `npm version [type]` beforehand if needed)
PKG_VER=$(cat package.json | grep version | head -1 | awk -F: '{ print $2 }' | sed 's/[",]//g')

# build as normal
npx tsc -p tsconfig.build.json
npm publish . --access=public

# build ESM
npx tsc -p tsconfig.build.json --module esnext # or whichever ESM
npm version --no-git-tag-version $PKG_VER-esm # different publish needs different version
npm publish. --access=public --tag esm # availiable at @typegoose/typegoose@esm

# go back to $PKG_VER from $PKG_VER-esm
npm version --no-git-tag-version $PKG_VER
```
Basically;
1. build & publish exactly as currently done
2. set version to have `-esm` suffix
3. build to ESM & publish to `@esm` tag
4. remove version `-esm` suffix

This way anyone who doesn't want ESM can continue as normal, because for them nothing has changed. Anyone who does want ESM simply installs the `@esm` tag, and can use ESM.

---

This is taken from a publish script I have for one of my own packages, which runs the following code;

```sh
# prebuild checks
npm run install
npm run prettier:check || exit 1
npm run test || exit 1
# get version
PKG_VER=$(cat package.json | grep version | head -1 | awk -F: '{ print $2 }' | sed 's/[",]//g')
# build & publish @latest (ESM)
npm run tsc:esm # package.json script for building ESM
npm publish . --access=public
# build & publish @esm (ESM)
npm version --no-git-tag-version $PKG_VER-esm
npm publish . --access=public --tag esm
# build & publish @cjs (CJS)
npm run tsc:cjs # package.json script for building CJS
npm version --no-git-tag-version $PKG_VER-cjs
npm publish . --access=public --tag cjs
# revert version & bump patch
npm version --no-git-tag-version $PKG_VER
npm version patch
```
As far as I've been able to test, there are no issues.

**Comment by hasezoey** [comment]
> Until you're ready to fully switch to ESM, a possibility might be to have an @esm tag?

i am not sure yet how it will be done, so i will first wait until i can even get ESM builds to work

> The build process might look like this (bash);

our build process is not just some bash / sh script, it uses `semantic-release`, though i dont think it will be much of a problem once actually going to do it

---

as a final note, typegoose will probably not switch fully to ESM until there are any benefits or mongoose switches to ESM (which will very likely not happen anytime soon)

**Comment by nhhockeyplayer** [comment]
10 months later and still waiting.... this is literally rendering this repo unusable
and the support thereof of this repo hairbrain
nothing is working for the field since introduction of ESM from angular and now that I have cleansed my back end of anything interfaces or anything hanging onto anything potentially angular dto, dao, entity... this crash continues

can we get esm on deck somehow?

**Comment by nhhockeyplayer** [comment]
well I think I seen enough of typegoose.... typeorm as well

heading to PRISMA

ciao

**Comment by nhhockeyplayer** [comment]
Im going to pull down the repo ands see what I can shake out for you here with esm

**Comment by nhhockeyplayer** [comment]
Hello committer, have you thought of this...

create a standard angular library as the new housing for @typegoose/typegoose
take the typegoose source code and dump it in
DONE

now try to ignore the word angular. What you would get is all the rigging and corner power you will need to produce a commercial class library in APF format that would have nothing to do with angular

most of your stuff in this scaffold would apply in some other tweeked form or another and likely work as-is like the work flow, all the config and test files to a predictable degree im sure a few gotcha's needing to be done differently

but this would secure building the library self contained in an esm centric scaffold

angular scaffolds come with alot of good stuff done right 

it would take 5mins to whip off an angular lib
then drop the typegoose code in

get started here
https://angular.io/guide/creating-libraries

```
ng new typegoose --no-create-application
cd typegoose
ng generate library typegoose
```

then you would edit the build files to reflect npm registry packaging and naming making sure you keep the project names de-coupled from the package names both live in the build space together and just need to be spelled out and kept separated to avoid confusion

- build
- pack (maybe for dev test)
- link (maybe for dev test)
- publish
DONE


I just got thru writing a professional lib for liberty mutual; a theme centric sass library using angular framework and the sass has nothing to do with angular.

end result was very moving.... 

typegoose can be same and I think esm would get nailed just by placing "type": "module" in the package.json and thats it

**Comment by nhhockeyplayer** [comment]
Additionally the same could be done using @nrwl/nx

now @nrwl/nx may be the better way to go since they are a wrapper to angular and optimize angular in many ways... its a better @angular/cli altogether

BUT as to why @nrwl/nx? because the project files for any library are de-coupled and you can add and operate properties such as "root" or "rootDir" and so forth that affect the code as it behaves within the library itself during consumption.

get started
`npm install -g nx`
https://nx.dev/getting-started/nx-setup
`npx create-nx-workspace@latest`
or preferably empty core scaffold
`npx create-nx-workspace --preset=core`

then add your lib
https://nx.dev/angular-tutorial/08-create-libs

`node node_modules/@nrwl/cli/bin/nx.js generate @nrwl/angular:library --name=typegoose  --publishable --buildable --prefix=typegoose --skipFormat --addModuleSpec  --tags='core,scope:shared,platform:web,type:orm' --unitTestRunner=jest --importPath=@typegoose/typegoose --strict --linter=eslint --standaloneConfig --compilationMode=partial --no-interactive --dry-run`

will generate this

```
CREATE libs/typegoose/README.md
CREATE libs/typegoose/ng-package.json
CREATE libs/typegoose/package.json
CREATE libs/typegoose/tsconfig.lib.json
CREATE libs/typegoose/tsconfig.lib.prod.json
CREATE libs/typegoose/tsconfig.spec.json
CREATE libs/typegoose/src/index.ts
CREATE libs/typegoose/src/lib/typegoose.module.ts
CREATE libs/typegoose/tsconfig.json
CREATE libs/typegoose/src/lib/typegoose.module.spec.ts
UPDATE tsconfig.base.json
CREATE libs/typegoose/jest.config.ts
CREATE libs/typegoose/src/test-setup.ts
CREATE libs/typegoose/.eslintrc.json
CREATE libs/typegoose/project.json
```

then just add the ESM type to package.json

I would go with @nrwl/nx to tinker with module search issues thereafter

**Comment by nhhockeyplayer** [comment]
here is my first stab at it
I have an Nx workspace created preset=core

just dropped all typoegoose sources inn it and it buiulds nicely an APF formatted lib

cmds follow... excuse the mix/match/do undo redo

```
1049  mkdir nx-typegoose
 1050  cd nx-typegoose
 1051  npx create-nx-workspace --preset=core\n\n
 1052  vi ~/.npmrc
 1053  npx create-nx-workspace --preset=core\n\n
 1054  ls
 1055  ls @typegoose
 1056  rm -rf *
 1057  npx create-nx-workspace --preset=core\n\n
 1058  ls
 1059  cd typegoose
 1060  ls
 1061  vi package.json
 1062  yarn
 1063  ls
 1064  yarn install
 1065  rm -rf node_modules
 1066  rm -rf node_modules
 1067  yarn install
 1068  npm run typegoose-migrate
 1069  npm run typegoose-migrate
 1070  npx nx format:write
 1071  npx nx format:check
 1072  node node_modules/@nrwl/cli/bin/nx.js generate @nrwl/angular:library --name=typegoose --publishable --buildable --prefix=typegoose --skipFormat --addModuleSpec --tags='domain:core,scope:shared,platform:node,type:orm' --unitTestRunner=jest --importPath=@typegoose/typegoose --strict --linter=eslint --standaloneConfig --compilationMode=partial --no-interactive --dry-run
 1073  yarn add @nrwl/angular
 1074  pwd
 1075  yarn add @nrwl/angular -W
 1076  node node_modules/@nrwl/cli/bin/nx.js generate @nrwl/angular:library --name=typegoose --publishable --buildable --prefix=typegoose --skipFormat --addModuleSpec --tags='domain:core,scope:shared,platform:node,type:orm' --unitTestRunner=jest --importPath=@typegoose/typegoose --strict --linter=eslint --standaloneConfig --compilationMode=partial --no-interactive --dry-run
 1077  node node_modules/@nrwl/cli/bin/nx.js generate @nrwl/angular:library --name=typegoose --publishable --buildable --prefix=typegoose --skipFormat --addModuleSpec --tags='domain:core,scope:shared,platform:node,type:orm' --unitTestRunner=jest --importPath=@typegoose/typegoose --strict --linter=eslint --standaloneConfig --compilationMode=partial --no-interactive 
 1078  node node_modules/@nrwl/cli/bin/nx.js generate @nrwl/workspace:remove typegoose
 1079  node node_modules/@nrwl/cli/bin/nx.js generate @nrwl/angular:library --name=typegoose --publishable --buildable --prefix=typegoose --skipFormat --addModuleSpec --tags='domain:core,scope:shared,platform:node,type:orm' --unitTestRunner=jest --importPath=@typegoose/typegoose --strict --linter=eslint --standaloneConfig --compilationMode=partial --no-interactive 
 1080  node node_modules/@nrwl/cli/bin/nx.js generate @nrwl/angular:remove typegoose
 1081  node node_modules/@nrwl/cli/bin/nx.js generate @nrwl/angular:library --name=typegoose --publishable --buildable --prefix=typegoose --skipFormat --addModuleSpec --tags='domain:core,scope:shared,platform:node,type:orm' --unitTestRunner=jest --importPath=@typegoose/typegoose --strict --linter=eslint --standaloneConfig --compilationMode=partial --no-interactive 
 1082  node node_modules/@nrwl/cli/bin/nx.js generate @nrwl/angular:remove typegoose
 1083  node node_modules/@nrwl/cli/bin/nx.js generate @nrwl/angular:library --name=typegoose --publishable --buildable --prefix=typegoose --skipFormat --addModuleSpec --tags='domain:core,scope:shared,platform:node,type:orm' --unitTestRunner=jest --importPath=@typegoose/typegoose --strict --linter=eslint --standaloneConfig --compilationMode=partial --no-interactive 
 1084  nx run-many --all --target=build
```

**Comment by nhhockeyplayer** [comment]
here is the repo if you want to tinker with it

https://github.com/nhhockeyplayer/nx-typegoose

**Comment by nhhockeyplayer** [comment]
so all of your .workflow stuff was just dropped in

you can pull this down and see if it flies in your registry space

I already put in the ESM format property in package.json

no test added just the core source

good luck

**Comment by nhhockeyplayer** [comment]
![Screen Shot 2022-07-26 at 6 40 59 PM](https://user-images.githubusercontent.com/6886551/181124561-b28c0a66-3481-42fa-9501-ca5a70a1a8cb.png)

**Comment by nhhockeyplayer** [comment]
let me know how you feel about this...

I think you have better clarity over the src code

and if this is acceptable

its all high velocity
easier repo maintenance

am happy to help you automate auto login for github actions and npmjs

but would like to see a new typegoose release

im oughta here

**Comment by nhhockeyplayer** [comment]
Im going to do an npm pack 
add the tgz to my package.json
and see if that gets me anywhere in my consuming project

**Comment by nhhockeyplayer** [comment]
[typegoose-0.0.0.tgz](https://github.com/typegoose/typegoose/files/9193765/typegoose-0.0.0.tgz)

**Comment by nhhockeyplayer** [comment]
[typegoose-typegoose-0.0.1.tgz](https://github.com/typegoose/typegoose/files/9193775/typegoose-typegoose-0.0.1.tgz)

**Comment by nhhockeyplayer** [comment]
well I couldnt resist... I redid the src code for better strict compilation passing
added beachball for your convenience
copied in your website
added nx migrate to sync dependencies to latest with 1 package.json script cmd
renamed barrel file to publis-api.ts
added output api to package.json for lib

other chores

pushed to branch

here is product
[typegoose-typegoose-0.0.1.tgz](https://github.com/typegoose/typegoose/files/9193930/typegoose-typegoose-0.0.1.tgz)

**Comment by nhhockeyplayer** [comment]
Just some codebase feedback...

Im modeling the ESM/APF right now for typegoose... (the package.json and its 2ndry entrypoint)

`lib/index.ts` should have been names indexes.ts to prevent collisions with barrel files and other frameworks... 
Im working this in now

the public api is hardwired to `src/index.ts` hence the 2ndry entrypoint for ng-packagr to get this into APF/ESM packaged up distribution

commits/push pending before eod today on same repo I mentioned above

I thought I would bring nestjs-typegoose along for the ride... cloned and getting that sync'd up... probably should fork it if this POC flies

no feedback... whats up?

consumption report pending

**Comment by nhhockeyplayer** [comment]
I did peek at the code for enums in typegoose...

enums have quirks in typescript... I always sought after an optimal usage because they are best practice and way to go if they are nailed properly and needed everywhere

but my first take it Im unsure if those are legit... would need to confirm once I get deeper into this

but my first glance is unsure

**Comment by nhhockeyplayer** [comment]
anyway I haven't forked anything and wish I had given all the changes I performed.

Im happy to follow thru with a fork if the committer is alright 
OR
let the committer run with what I POC'd

results pending

**Comment by nhhockeyplayer** [comment]
the nestjs-typegoose distribution is a mess... as its currently released.
this could very easily be cleaned up and automated with the same @nrwl/nx library scaffolding for APF/ESM I am exercising now and would be the way to get this guy on deck for ESM. Im reverting back to the released version of nestjs-typegoose and will think about dealing with this after typegoose.

so far looking good about to build...

drum roll please

**Comment by AndyClausen** [comment]
Dude, every time you comment you're notifying like 20 people.

It's great you want to help, but could you open a PR draft linked to this issue and comment there instead? Not everyone here is as invested in this issue.

**Comment by nhhockeyplayer** [comment]
Well my knowledge of the internals are off.

I have my consuming project (meanstacknh) consuming my new nx-typegoose and nx-nestjs-typegoose (anchored to nx-typegoose)

But my consuming project is barking during compilation and Im being blocked from qualifying this further due to the following lines of code within typegoose/internal/utils.ts

Utils.ts:478
```
  if (Type instanceof mongoose.Schema) {
    // error TS2322: Type 'SchemaTypeOptions<any>' is not assignable to type 'AnyParamConstructor<any>'.
    OptionsCTOR = mongoose.Schema.Types.Subdocument.prototype.OptionsConstructor;
  }
```

and im blue in the face cannot get this to articulate so the consuming project can compile


```
this is the actual error during build on the consuming project (meanstacknh)

        >  NX   node_modules/@typegoose/typegoose/lib/internal/utils.ts:471:5 - error TS2322: Type 'SchemaTypeOptions<any>' is not assignable to type 'AnyParamConstructor<any>'.
       
            Type 'SchemaTypeOptions<any>' provides no match for the signature 'new (...args: any): any'.
          
          471     OptionsCTOR = mongoose.Schema.Types.Subdocument.prototype.OptionsConstructor
                  ~~~~~~~~~~~
```

if anyone can offer a push in the right direction for this I would appreciate that.
I dont want to goto prisma but Im almost there
would rather stay with typegoose

thanks for any assist

**Comment by hasezoey** [comment]
i have updated the `feature/esm` branch again, this time almost everything just works, but is currently blocked by:
- [ ] https://github.com/facebook/jest/issues/11644

**Comment by nhhockeyplayer** [comment]
Thanks committer for gaining traction this is very exciting.

I have demonstrated `@nrwl/nx` as THE platform by which typegoose should be and appropriately be managed and modeled its the silver bullet to all dependencies high velocity wraps jest as well and lots more a comprehensive solution.

If your running into those jest issues I would highly encourage you take a peek at the repo's I modeled for you and seriously consider platforming typegoose on `@nrwl/nx` . The @nrwl team are the marines when it comes to resolving open issues with workaround and patches under their hood to circumvent and resolve open issues jest, angular, webpack and other dependent frameworks fail to address... their footprint would not survive if they did not make everything fly even when its broken with real world workarounds and patches.

I got typegoose and nestjs-typegoose platformed in @nrwl/nx in just a few hours its command centric and schematic driven.

Trying this flavor of building software I know you will be very impressed. Makes everything easier and like I said it may resolve any jest test issues your encountering its worth investigating and attempting.

I will give your esm branch a try (feedback pending)

Just getting things on the map without test is worth it all.
Im always here to help if your going to attempt platforming typegoose on `@nrwl/nx`
Its common sens, typegoose will rule the future no doubt if you make that leap.

**Comment by nhhockeyplayer** [comment]
Wish I could stay on but Im dropping my typegoose models for graphql models

new things are emerging like fauna modeling natural relationships with models.

Good luck with this seems like 90% there but 2 yrs and waiting too long

if typegoose nails this anytime soon days/weeks I might look back but Im beginning my transition

graphql though is trending in the field

- cheers

**Comment by smolinari** [comment]
@nhhockeyplayer - GraphQL has absolutely nothing to do with which database you have storing what data. In fact, one of the problems GraphQL solves is the need for a Gateway to be able to collect data from a plethora of different data sources. In other words, "GraphQL Models" are about the data sent back and forth via the API. The goal of these models is to form the data shapes needed by the UI, which is another one of its strengths. This in turn means absolutely nothing about the data models needed to make business decisions or to persist state for your app(s) (i.e. the data in databases). I believe, if you continue with the mental model you've expressed above about GraphQL, you'll run into a lot of issues. Just trying to hopefully avoid that for you.  

Fauna is an interesting NoSQL database for sure, but it is a DBaaS. That means, to use it, you have to take what they offer. You can't build your own database server with it. That can mean two things, 1. you'll have to pay a good penny to use it professionally. 2. you'll more than likely have network latentencies to contend with, depending on where you place your app. Maybe neither are a problem for you. Just sayin. 

I didn't check this, but I'd bet you can have different package module resolution scenarios between your apps with Nx. I'd be a really piss-poor monorepo manager if couldn't. If it doesn't, I can suggest using Rush. It can. I know, because I have this going with it right now. 

At any rate, I hope you find success no matter what you end up doing. 

Scott

**Comment by nhhockeyplayer** [comment]
Can anyone shed light on this issue since the 4 yrs of its citing?

ESModules

Can I use typegoose with my favorite frameworks now or is it still limited and stuck

any movement on this?
has this ticket gone stale and a non-issue

is typegoose going anywhere?

is anybody home?

**Comment by hasezoey** [comment]
@nhhockeyplayer please read the thread.
The last time i tried i ran into issues, most notably jest (which is the testing framework we use and i dont want to switch) didnt support ESM [without a experimental nodejs flag](https://jestjs.io/docs/ecmascript-modules) (which would mean all users of typegoose would also have to use that flag if jest is used), there were some additional jest issues (see https://github.com/jestjs/jest/issues/11644 for example).
Aside from that i had also some issues with typescript, but they have improved ESM handling since then. (though note that they still have no replacement for `emitDecoratorMetadata` for ESM decorator metadata)

Finally, if your issue is still "which compels @nrwl/nx and they mandate everything be ESM", it will be a long time since mongoose is still CJS (though providing [ESM-compatible  imports](https://github.com/Automattic/mongoose/blob/2607904171d5acacd468845083f88a806e032ff6/index.js#L14-L64)).

**Comment by hasezoey** [comment]
Once again because of interest, i have updated the [`feature/esm`](https://github.com/typegoose/typegoose/tree/feature/esm) branch, which has a `build`(cjs) and `build:esm`(mjs) options and can be a hybrid package, though because of the issue with jest, the ESM cannot be tested and so can only be manually tested.
I have tested it to work via [typegoose-testing](https://github.com/typegoose/typegoose-testing/tree/test_esm).

Some issues / annoyances i have encountered this time:
- jest still does not support ESM without a [experimental nodejs flag](https://jestjs.io/docs/ecmascript-modules)
- typescript requires setting top-level `package.json` `type: module` to build any ESM at all (even if the tsconfig says `module: node16`), which in turn:
- requires [stub-package.json's](https://github.com/typegoose/typegoose/blob/35c8bfe1cff66d6fbc237a7d2fb00e0079d98964/stubs/cjs-package.json) to define what each type is (otherwise the `lib/typegoose.js`, defined as cjs, will try to load as ESM for whatever reason)
- using `type: module` requires enabling `esModuleInterop`, which in turn makes some imports require to use a different style (use `import lodash from ...` instead of `import * as lodash from`), without any errors from typescript and only runtime nodejs errors

**Comment by nhhockeyplayer** [comment]
great

if you can release it with those quarks outstanding

I will give typegoose a run and see how she drives

can we go that route?

Im not concerned with testing like a formidable corporation would

but if I can operate the released package given a couple quarks thats fine with me
I just hope its not interrupting to development

I just want to download the package and run with it

can we go this route since its been several years?

my feedback will be helpful in the coming weeks

**Comment by nhhockeyplayer** [comment]
I have a formidable api backend from 2 yrs ago I aborted due to this esm issue

its heavily mechanized with typegoose models

I built my own layer of intrinsic deep seeding controllers to boot and a hierarchy of services

all tied into postgres previously serving up REST friendly json

but my main goal now is to get the back end api to speak graphql
and start tossing around graphql models served up by typegoose models from the back end

Im not fresh on my backend api I built its coded to https://github.com/m8a-io/m8a
was this the solution workaround for this ticket?

if anyone can clarify for me to get me back to base I would be greatful

my code uses postgres but the connect string has mongo and if mongo is not esm compliant Im blocked?

Im still wondering why m8 didnt work for me and Im guessing the esm issue

not sure what my options are at this point

```
import { ConfigService } from '@nestjs/config';
import { TypegooseModuleOptions } from '@m8a/nestjs-typegoose';

export const getMongoConfig = async (
    configService: ConfigService
): Promise<TypegooseModuleOptions> => {
    return {
        uri: getMongoString(configService),
        ...getMongoOptions(),
    };
};

const getMongoString = (configService: ConfigService) =>
    `mongodb://${configService.get('MONGO_HOST')}:${configService.get(
        'MONGO_PORT'
    )}/${configService.get('MONGO_AUTHDATABASE')}`;
// const getMongoString = (configService: ConfigService) => 'mongodb://localhost:27017/energy';

const getMongoOptions = () => ({
    useNewUrlParser: true,
    useUnifiedTopology: true,
});
```

**Comment by nhhockeyplayer** [comment]
this is what I get when I launch my backend admin-api based on m8/nestjs-typegoose

but I cant for the life of me remember why I quit developing it

it limps to a point

postgres upgraded and running

mongo is a curveball for me though the code is anchored to types but I dont think Im using a mongodb which should free me from the esm issue... 

still trying to figure

```
meanstack02@admin meanstacknh % nx serve admin-api --verbose
[NX CLOUD] Using light client
[NX CLOUD] A local bundle currently exists:  {
  version: '2412.05.4',
  fullPath: '/Users/meanstack02/meanstacknh/node_modules/.cache/nx/cloud/2412.05.4'
}
[NX CLOUD] Last verification was more than 30 minutes ago, verifying bundle is still valid
[NX CLOUD] Currently installed bundle is invalid, downloading version 2412.05.4  from  https://cloud.nx.app/nx-cloud/static/client-bundle
[NX CLOUD] Using bundle path:  {
  version: '2412.05.4',
  fullPath: '/Users/meanstack02/meanstacknh/node_modules/.cache/nx/cloud/2412.05.4'
}
Was not able to require.resolve module nx/src/utils/app-root from the following paths: /Users/meanstack02/meanstacknh/node_modules/nx-cloud/node_modules,/Users/meanstack02/meanstacknh/node_modules. This may be expected.
Was not able to require.resolve module nx/src/utils/app-root from the following paths: /Users/meanstack02/meanstacknh/node_modules/nx-cloud/node_modules,/Users/meanstack02/meanstacknh/node_modules. This may be expected.
Nx is configured to use the legacy cache. This cache will be removed in Nx 21. Read more at https://nx.dev/deprecated/legacy-cache.
[Nx Cloud Debug] Attempting to acquire filesystem lock with path:  /var/folders/73/pgw9tfdj5fz826zd6cc4p0vr0000gn/T/client-instance-id.lock
[Nx Cloud Debug] Failed to create folder lock at path: /var/folders/73/pgw9tfdj5fz826zd6cc4p0vr0000gn/T/client-instance-id.lock
[Nx Cloud] Unable to detect a VCS context from the environment.

 NX   RunStart


{
  "meta": {},
  "branch": null,
  "runGroup": "60c7417435d178a7f7fd02d9368ff0148edcf99e",
  "ciExecutionId": null,
  "ciExecutionEnv": "",
  "hashes": [
    "16086696159647929764"
  ],
  "machineInfo": {
    "machineId": "uHiOzR0l5SEx0kjZqAemHQ==",
    "platform": "darwin",
    "version": "Darwin Kernel Version 21.6.0: Mon Jun 24 00:56:10 PDT 2024; root:xnu-8020.240.18.709.2~1/RELEASE_X86_64",
    "cpuCores": 8
  },
  "vcsContext": null,
  "clientInstanceSource": "CLOUD_RUNNER",
  "clientInstanceId": "f8f4420a-03ac-4f6d-b05e-bcf788c41590"
}


> nx run admin-api:serve

RunStart duration: 104
                      (node:55458) [DEP0180] DeprecationWarning: fs.Stats constructor is deprecated.
(Use `node --trace-deprecation ...` to show where the warning was created)
asset main.js 40.3 KiB [emitted] (name: main)
asset assets/.gitkeep 0 bytes [emitted] [from: src/assets/.gitkeep] [copied]
  
webpack 5.96.1 compiled successfully (a2db65b57797a851)
Debugger listening on ws://localhost:9229/5f3bf6e6-6dd2-4d9a-9df3-3fc054ad1a28
Debugger listening on ws://localhost:9229/5f3bf6e6-6dd2-4d9a-9df3-3fc054ad1a28
For help, see: https://nodejs.org/en/docs/inspector

[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [NestFactory] Starting Nest application...
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [InstanceLoader] TypegooseModule dependencies initialized +20ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [InstanceLoader] PassportModule dependencies initialized +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [InstanceLoader] ConfigModule dependencies initialized +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [InstanceLoader] ConfigModule dependencies initialized +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [InstanceLoader] AppModule dependencies initialized +2ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [InstanceLoader] JwtModule dependencies initialized +16ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [InstanceLoader] TypegooseCoreModule dependencies initialized +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [InstanceLoader] TypegooseModule dependencies initialized +15ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [InstanceLoader] AuthModule dependencies initialized +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [InstanceLoader] UserModule dependencies initialized +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RoutesResolver] AppController {/nest}: +30ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RouterExplorer] Mapped {/nest, GET} route +2ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RoutesResolver] AuthController {/nest/auth}: +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RouterExplorer] Mapped {/nest/auth/register, POST} route +1ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RouterExplorer] Mapped {/nest/auth/login, POST} route +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RoutesResolver] UserController {/nest/user}: +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RouterExplorer] Mapped {/nest/user/:id, GET} route +1ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RouterExplorer] Mapped {/nest/user, GET} route +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RouterExplorer] Mapped {/nest/user, POST} route +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RouterExplorer] Mapped {/nest/user/bulk, POST} route +1ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RouterExplorer] Mapped {/nest/user/:id, PATCH} route +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RouterExplorer] Mapped {/nest/user/:id, PUT} route +0ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [RouterExplorer] Mapped {/nest/user/:id, DELETE} route +1ms
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG [NestApplication] Nest application successfully started +2ms
(node:55463) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)

[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG main.ts Nest API Listening at http://localhost:3333/nest
[Nest] 55463  - 12/10/2024, 2:27:38 AM     LOG main.ts Mongoose TYPEGOOSE database Listening at mongodb://127.0.0.1:27017/energy


/Users/meanstack02/meanstacknh/node_modules/mongoose/lib/connection.js:825
  const serverSelectionError = new ServerSelectionError();
                               ^
MongooseServerSelectionError: connect ECONNREFUSED 0.0.0.0:27017
    at NativeConnection.Connection.openUri (/Users/meanstack02/meanstacknh/node_modules/mongoose/lib/connection.js:825:32)
    at Mongoose.createConnection (/Users/meanstack02/meanstacknh/node_modules/mongoose/lib/index.js:356:10)
    at InstanceWrapper.useFactory [as metatype] (/Users/meanstack02/meanstacknh/node_modules/@m8a/nestjs-typegoose/dist/typegoose-core.module.js:58:43)
    at Injector.instantiateClass (/Users/meanstack02/meanstacknh/node_modules/@nestjs/core/injector/injector.js:376:55)
    at callback (/Users/meanstack02/meanstacknh/node_modules/@nestjs/core/injector/injector.js:65:45)
    at Injector.resolveConstructorParams (/Users/meanstack02/meanstacknh/node_modules/@nestjs/core/injector/injector.js:145:24)
    at Injector.loadInstance (/Users/meanstack02/meanstacknh/node_modules/@nestjs/core/injector/injector.js:70:13)
    at Injector.loadProvider (/Users/meanstack02/meanstacknh/node_modules/@nestjs/core/injector/injector.js:98:9)
    at /Users/meanstack02/meanstacknh/node_modules/@nestjs/core/injector/instance-loader.js:56:13
    at async Promise.all (index 4)


 NX  Process exited with code 1, waiting for changes to restart...
```

**Comment by nhhockeyplayer** [comment]
Thanks Scott... appreciate your generosity.
I had to take a nap on what I was doing back then.
Im prejudice against promises. And built my own service layer that supports promises and observables for everything.
But I liked your bootstrap configuration. So thats why my code is like this wired into your stuff but unfinished.
My. ain issue was my models are isomorphic (shared with back end and front end) and this is where the esm issue comes into play my models use mongo types like ID and when nx/angular try to build them it quits on esm failure.
This is why I would like to see if typegoose can release with the esm branch and my formidable setup can give it a hard run.

In the interim until hazeohey give the nod

Im going to troubleshoot this connection issue

but I need a back end badly Im tired of json-server and json-graphql-server
and need to do auth normally

waiting on the good news
- cheers

**Comment by nhhockeyplayer** [comment]
I did have a question about models

should they be parked on the front end or the back end or shared

my issue was they were shared so the front end could stay synchronized

**Comment by hasezoey** [comment]
> I did have a question about models

This question is off-topic and i will mark it as such, but because you asked it anyway:

> my issue was they were shared so the front end could stay synchronized

My personal (and what seems many others) recommended way is to not share the Database Model with the front-end and do a service layer in-between, this way you could do some more validation, transforming or even change / update the database layer without having to deal with front-end issues (and likely other problems), though yes this would initially create a bunch of boilerplate.
Additionally, while mongoose *may* work in the front-end (without a driver) it is to my knowledge still not a good way to actually run it (and typegoose may also work there, but i would still call that unsupported, but will try to accommodate anyone having issues there).
See #33 for more talk regarding typegoose on the client side.

**Comment by nhhockeyplayer** [comment]
Typegoose is a "wrapper" for easily writing Mongoose models with TypeScript.

why cant typegoose be a wrapper for easily writing PostGRES models with typescript

**Comment by hasezoey** [comment]
If you have further off-topic question, please open [a new Question discussion](https://github.com/typegoose/typegoose/discussions/categories/q-a).

> why cant typegoose be a wrapper for easily writing PostGRES models with typescript

Because typegoose is (at least currently, and likely for the future) closely tied to the mongoose schema & model structure, not even to speak about the types.

**Comment by nhhockeyplayer** [comment]
well for mongo to be esm noncompliant is really an impediment

hoping for a way around that

Nx has really taken off and I use it at very high velocity
and for being unable to even compile for a type ID is unspeakable impediment
the reviews of mongo for 2024 are hugely positive

**Comment by hasezoey** [comment]
Updated `feature/esm` once again for typegoose 12.10.0 (mongoose 8.9.0), where some patches that could be applied (from `feature/esm`) were upstreamed to `master` (for 12.10.0)

**Comment by nhhockeyplayer** [comment]
replying to your response in this ticket
https://github.com/typegoose/typegoose/issues/972

Nx is turnkey
It can do one single lib
or
it can do the entire world

one workspace
one library
built in support for elegant github publishing (just a few npm commands in package.json)

I will sum up the transition in just a few terminal commands

Nx is not NEXT (next is something else)
Its tragic you dont know about Nx 
It cant be caught 
was way ahead of its time in 2019
world has caught on too
very ripe

Nx is @nrwl/nx and there is no framework like it on the planet high velocity, turnkey, does everything right first time.

docs for esm library generator https://nx.dev/nx-api/js/generators/library (it will create esmodule lib)

okay here goes to kick start a repo for @typegoose/typegoose

open a mac terminal
create-nx-workspace  is where to begin
`npx create-nx-workspace --help`
peek at option `--preset`
I think the preset you want it `ts`

1. npx create-nx-workspace --preset=ts

meanstack02@admin 2025 % npx create-nx-workspace --preset=ts

```
 NX   Let's create a new workspace [https://nx.dev/getting-started/intro]

✔ Where would you like to create your workspace? · typegoose
✔ Would you like to use Prettier for code formatting? · Yes
✔ Which CI provider would you like to use? · github

 NX   Creating your v20.1.0 workspace.

✔ Installing dependencies with yarn
⠼ Creating your workspace in typegoose.
✔ Nx Cloud has been set up successfully
✔ CI workflow has been generated successfully

 NX   Your CI setup is almost complete.

Finish it by visiting: https://cloud.nx.app/connect/Fh1zTGH4d4
```

and theres your workspace/monorepo open it up with your IDE and goto next terminal window to create lib

now lets cd into that dir with terminal or just open it up and launch terminal in the root project dir

2. cd typegoose
3. at this time you should be already logged into npm for publishing in the terminal your working
4. create your library using nx generator for js

`nx g @nx/js:library --directory=libs/typegoose --bundler=esbuild --importPath=@typegoose/typegoose --linter=eslint --name=typegoose --publishable=true --unitTestRunner=jest --includeBabelRc=true`

now just copy your typescript code from original lib into libs/typegoose/src/lib
dont forget to generate a new barrel file or public-api.ts whichever you like

here is the build command to build it all

`nx run-many --all --target=build` from root project dir

add a few scripts to package.json for pack & maintenance (not debugged yet)
```
"publish-internationalization": "cd libs/typegoose/dist && npm publish",
"pack-internationalization": "cd libs/typegoose/dist && npm pack",
```
1st time for me its a little different than integrated monorepo so you may need to work this out debug running it


DONE

esm compliant
turnkey coding and building and deploying
support for everything
doesn't get any easier or sharper 
Nx does it all for you

and you might wet your pants its so elite

I really hope you take a bite out of Nx and get typegoose where it belongs
at the top

**Comment by nhhockeyplayer** [comment]
eagerly awaiting feedback on this

typegoose cannot lose

jest full support

everything works

faster
easier
more dynamic and fun

**Comment by nhhockeyplayer** [comment]
NxCloud is a feature of Nx
just create an account there (make sure it is the Nx team one)
it will automagically connect your repo to their build cloud (no need to be logged in its under the hood perf boost)
and cache all your tasks being built
for lighting fast build speed

just a side show but its fun

and tons of github/github-actions support automated

**Comment by nhhockeyplayer** [comment]
just a final word

I use nx-console plugin in the IDE to pre-qualify all nx generator options with --dry-run
nothing gets written unless your pin point precise
remove --dry-run and it generates all your code
```
e2e
jest
eslint
everything you want or need optionally

![Image](https://github.com/user-attachments/assets/e8b76e51-ebdb-4f80-9106-3f1cb79dd5e4)

```

I created an Nx scaffold for typegoose in 10mins

I really hope you do this could have been done back in 2019

- cheers

**Comment by nhhockeyplayer** [comment]
well I kind of ran ahead and did my own typegoose lib according to instructions above

realized i screwed up
should have located the lib under packages instead of libs on generation

I was able to compile all of typegoose though and get it building

but since angular is esm compliant since angular 17 and typegoose can be construed as isomorphic, it dont matter whether its front end or backend.. its isomorphic
just esm models

and angular library generator is air tight I already have libraries published in my own personal capacity that are razor precise

so Im going to recommend using the angular library generator for typegoose
why?
the JS generator may be missing some packaging implementation automation (its always a moving target and never know whats actually ripe)
angular we know is ripe
but sing that generator is optional
if the JS generator has issues I will opt to use angular APF esm compliant format its fairly resilient

but for now Im going to complete this one for packages with the JS generator
and run ahead and see how it works published out to my github
and pulled down for my mobile app and nestjs server

results pending

dont wait for me go ahead and place typegoose into an Nx scaffold

**Comment by nhhockeyplayer** [comment]
I assert there has got to be some hoisting issues that remain automated by Nx at this time
https://github.com/nrwl/nx/issues/29937

so I will be trying all generators

- @nx/node
- @nx/js
- @nx/angular
- @nx/webpack
- @nx/esbuild
- @nx/web
- @nx/nest

to try and find out what works and what can be integrated elegantly and coherent

goal: create typegoose library that compiles and runs with esm frameworks

everything is esm now anyway

and these backend issues need to be eradicated so fluid UI Development can resume

**Comment by nhhockeyplayer** [comment]
progress update
I did create my own non-publishable APF format typegoose
buildable against "types":["node"] inside my tsconfig.lib.json

builds against all my middle tier abstract and isomorphic classes that all reference and build against mongoose, nestjs and typegoose

havent run anything but Im confident, will get to that

In any event just to say I created this using Nx

attached is the project zip

the Nx issues I cited past week reflect what I feel are circular dependencies within my large workspace and have nothing to do with typegoose buildability or runnability and when Nx gets their dep-graph back to par I should thresh those out

[typegoose.zip](https://github.com/user-attachments/files/18721695/typegoose.zip)

**Comment by nhhockeyplayer** [comment]
Hi Folks,

Please excuse my last post

Nx has not cleared esm compliance as of yet.

Working it

You can hold on esm with Nx until I give go ahead.

Theory is there

Need to confirm runtime

thanks for understanding

**Comment by nhhockeyplayer** [comment]
Hi Typegoose Committer(s)

Im finished with my Nx issues. Circular dependencies.

I built a @typegoose/typegoose within my Nx scaffold using esbuild nx generator

Below are the results

I will watch for any pending releases to sync up with the real repository.

For now Im moving on


[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [NestFactory] Starting Nest application...
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] MongooseModule dependencies initialized +17ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] AppModule dependencies initialized +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] ConfigModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] ConfigModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] UserModule dependencies initialized +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] PermissionModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] RolePermissionModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] RoleModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] JwtModule dependencies initialized +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [NestFactory] Starting Nest application... +39ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] MongooseModule dependencies initialized +5ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] RolePermissionSeedModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] PermissionSeedModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] RoleSeedModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] UserSeedModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] ConfigModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] ConfigModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] IntrinsicSeedModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] RoleModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] PermissionModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] RolePermissionModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] UserModule dependencies initialized +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [InstanceLoader] JwtModule dependencies initialized +1ms
hey hi
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RoutesResolver] AppController {/nest}: +15ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest, GET} route +3ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RoutesResolver] UserController {/nest/user}: +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/user/:id, GET} route +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/user, GET} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/user, POST} route +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/user/bulk, POST} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/user/:id, PATCH} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/user/:id, PUT} route +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/user/:id, DELETE} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RoutesResolver] PermissionController {/nest/permission}: +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/permission/:id, GET} route +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/permission/:id, GET} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/permission, GET} route +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/permission, POST} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/permission/bulk, POST} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/permission/:id, PATCH} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/permission/:id, PUT} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/permission/:id, DELETE} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RoutesResolver] RolePermissionController {/nest/role_permission}: +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role_permission/:id, GET} route +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role_permission, GET} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role_permission, POST} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role_permission/bulk, POST} route +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role_permission/:id, PATCH} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role_permission/:id, PUT} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role_permission/:id, DELETE} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RoutesResolver] RoleController {/nest/role}: +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role/:id, GET} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role/:id, GET} route +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role, GET} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role, POST} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role/bulk, POST} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role/:id, PATCH} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role/:id, PUT} route +0ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [RouterExplorer] Mapped {/nest/role/:id, DELETE} route +1ms
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG [NestApplication] Nest application successfully started +1ms
hey hi
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG main.ts Nest API Listening at http://localhost:3333/nest
[Nest] 99483  - 02/12/2025, 6:56:25 PM     LOG main.ts Mongoose TYPEGOOSE database Listening at mongodb://127.0.0.1:27017/energy
Mongoose: role.insertOne({ name: 'ADMINISTRATOR', description: 'ADMINISTRATOR', isSuper: true, _id: ObjectId("67ad3529c9f2e04cf9b2fd98"), __v: 0}, {})
Mongoose: role.insertOne({ name: 'USER', description: 'USER', isSuper: false, _id: ObjectId("67ad3529c9f2e04cf9b2fd99"), __v: 0}, {})
Mongoose: role.insertOne({ name: 'MANAGER', description: 'MANAGER', isSuper: false, _id: ObjectId("67ad3529c9f2e04cf9b2fd9a"), __v: 0}, {})
Mongoose: role.insertOne({ name: 'IT', description: 'IT', isSuper: false, _id: ObjectId("67ad3529c9f2e04cf9b2fd9b"), __v: 0}, {})
Mongoose: role.insertOne({ name: 'STAFF', description: 'STAFF', isSuper: false, _id: ObjectId("67ad3529c9f2e04cf9b2fd9c"), __v: 0}, {})
Mongoose: role.insertOne({ name: 'OWNER', description: 'OWNER', isSuper: false, _id: ObjectId("67ad3529c9f2e04cf9b2fd9d"), __v: 0}, {})
Mongoose: role.insertOne({ name: 'MARKETER', description: 'MARKETER', isSuper: false, _id: ObjectId("67ad3529c9f2e04cf9b2fd9e"), __v: 0}, {})
Mongoose: role.insertOne({ name: 'ANONYMOUS', description: 'ANONYMOUS', isSuper: false, _id: ObjectId("67ad3529c9f2e04cf9b2fd9f"), __v: 0}, {})
Mongoose: role.insertOne({ name: 'SUBJECT', description: 'SUBJECT', isSuper: false, _id: ObjectId("67ad3529c9f2e04cf9b2fda0"), __v: 0}, {})
Mongoose: role.insertOne({ name: 'FRONTDESK', description: 'FRONTDESK', isSuper: false, _id: ObjectId("67ad3529c9f2e04cf9b2fda1"), __v: 0}, {})
Mongoose: role.insertOne({ name: 'MAINTENANCE', description: 'MAINTENANCE', isSuper: false, _id: ObjectId("67ad3529c9f2e04cf9b2fda2"), __v: 0}, {})
Mongoose: role.insertOne({ name: 'QA', description: 'QA', isSuper: false, _id: ObjectId("67ad3529c9f2e04cf9b2fda3"), __v: 0}, {})
Mongoose: permission.insertOne({ name: 'CREATE', description: 'CREATE', _id: ObjectId("67ad3529c9f2e04cf9b2fdaa"), __v: 0}, {})
Mongoose: permission.insertOne({ name: 'READ', description: 'READ', _id: ObjectId("67ad3529c9f2e04cf9b2fdab"), __v: 0}, {})
Mongoose: permission.insertOne({ name: 'UPDATE', description: 'UPDATE', _id: ObjectId("67ad3529c9f2e04cf9b2fdac"), __v: 0}, {})
Mongoose: permission.insertOne({ name: 'DELETE', description: 'DELETE', _id: ObjectId("67ad3529c9f2e04cf9b2fdad"), __v: 0}, {})
Mongoose: user.insertOne({ emailAddress: 'santa.clause@gmail.com', password: 'santa', role: ObjectId("67ad3529c9f2e04cf9b2fda4"), _id: ObjectId("67ad3529c9f2e04cf9b2fda5"), __v: 0}, {})
Mongoose: user.insertOne({ emailAddress: 'elf.clause@gmail.com', password: 'elf', role: ObjectId("67ad3529c9f2e04cf9b2fda6"), _id: ObjectId("67ad3529c9f2e04cf9b2fda7"), __v: 0}, {})
Mongoose: user.insertOne({ emailAddress: 'reindeer.clause@gmail.com', password: 'reindeer', role: ObjectId("67ad3529c9f2e04cf9b2fda8"), _id: ObjectId("67ad3529c9f2e04cf9b2fda9"), __v: 0}, {})
Mongoose: role.createIndex({ name: 1 }, { unique: true, background: true })
Mongoose: permission.createIndex({ name: 1 }, { unique: true, background: true })
Mongoose: card.createIndex({ cardNumber: 1 }, { unique: true, background: true })

**Comment by nhhockeyplayer** [comment]
A core packages monorepo that is scaffolded to accomodate all formats 'commonjs', 'esm', etc... all at once so when published all bases are covered is fairly turnkey

https://youtu.be/Vy4d0-SF5cY

doesnt get any simpler
